### PR TITLE
Assignment5

### PR DESCRIPTION
**Pull request by zander8-uw**

### CODE REVIEW

File: HuntTheWumpus/HuntTheWumpus.cpp [hunk 1]
```diff
@@ -201,66 +200,90 @@ int main()
...
-            if (stringTokens.size() < 2)
+            std::getline(std::cin, input);
+            // Split into strings.
+            const auto stringTokens = SplitString(input, " \t\n");
+
+            if (stringTokens.size() == 0)
             {
-                std::cout << "A Move command must be followed by the destination cave id.\n";
-                continue;
+                throw std::runtime_error("");
```
Comment (kbradl-uw) [inline HuntTheWumpus/HuntTheWumpus.cpp:217]: Since we're in a while loop, you can also realize the control flow change you want with a continue operation like on line 227.  Think then about why a run-time error is appropriate here vs. just messaging and continuing.

File: HuntTheWumpus/HuntTheWumpus.cpp [hunk 2]
```diff
@@ -201,66 +200,90 @@ int main()
...
-            const auto destCave = std::stoi(stringTokens[1]);
-
-            dungeon.MakeMove(HuntTheWumpus::DungeonMove::Move, { destCave });
-        }
+            const auto command = ToLower(stringTokens[0]);
 
-        if (command[0] == 's')
-        {
-            if (stringTokens.size() < 2)
+            if (command[0] == 'm' && command.length() != 0)
```
Comment (kbradl-uw) [inline HuntTheWumpus/HuntTheWumpus.cpp]: Your logic is inverted. If we access an empty command (and it's not obvious how we can take a string that's not whitespace and end up with an empty token) then you need to check that it's not empty before dereferencing the 0th element.

It's also better to use .empty() than .length() != 0 or .size() != 0.

```
if(command.empty())
{
    std::cout << "Could not understand the command " << input << std::endl;
    continue;
}

if(command[0] == 'm') 
```

Think also about how you could possible manifest command.length() == 0 or command.empty(). What possible input could you generate that would cause  this?

File: HuntTheWumpus/HuntTheWumpus.cpp [hunk 3]
```diff
@@ -201,66 +200,90 @@ int main()
...
-            // Remaining tokens is the desired arrow path.
-            std::vector<int> path;
-            auto firstToken = false;
+                // Second token is a destination.
+                const auto destCave = std::stoi(stringTokens[1]);
 
-            for (auto&& token : stringTokens)
+                dungeon.MakeMove(HuntTheWumpus::DungeonMove::Move, { destCave });
+            }
+            else if (command[0] == 's')
```
Comment (kbradl-uw) [inline HuntTheWumpus/HuntTheWumpus.cpp:236]: See above. If command is empty your if block above fails to be true, so we progress to this "else" block, where we then fail on dereferencing an empty string.

File: HuntTheWumpus/HuntTheWumpus.cpp [hunk 4]
```diff
@@ -201,66 +200,90 @@ int main()
...
-        }
+                    path.push_back(std::stoi(token));
+                }
+
+                path.resize(std::min(path.size(), PathLimit));
 
-        if (command[0] == 'q' || command[0] == 'e' || command[0] == 'x')
+                dungeon.MakeMove(HuntTheWumpus::DungeonMove::Shoot, path);
+            }
+            else if (command[0] == 'q' || command[0] == 'e' || command[0] == 'x')
```
Comment (kbradl-uw) [inline HuntTheWumpus/HuntTheWumpus.cpp:263]: Similar here.

File: HuntTheWumpus/HuntTheWumpus.cpp [hunk 5]
```diff
@@ -201,66 +200,90 @@ int main()
...
+            {
+                std::cout << "Exiting.\n";
+                change.GameOver(false);
+            }
+            else
+            {
+                throw std::invalid_argument("");
```
Comment (kbradl-uw) [inline HuntTheWumpus/HuntTheWumpus.cpp]: Consider the control flow here. You're throwing, which is a flow-control to the catch block, which is effectively the next line.

I suspect you can have the logic you've got for std::invalid_argument in line instead of in a catch block.

File: HuntTheWumpus/HuntTheWumpus.cpp [hunk 6]
```diff
@@ -201,66 +200,90 @@ int main()
...
+                    std::cout << "A Move command must be followed by the destination cave id.\n";
+                    continue;
+                }
+
 
-            // Remaining tokens is the desired arrow path.
-            std::vector<int> path;
-            auto firstToken = false;
+                // Second token is a destination.
+                const auto destCave = std::stoi(stringTokens[1]);
```
Comment (kbradl-uw) [inline HuntTheWumpus/HuntTheWumpus.cpp:232]: Can this fail in a way that's not addressed or be confused with a different recovery?

File: HuntTheWumpusLib/Dungeon.cpp [hunk 7]
```diff
@@ -157,6 +158,21 @@ namespace HuntTheWumpus
...
     {
+        // Guard against an empty destinationIds vector
+        try
+        {
+            if (destinationIds.empty())
+            {
+                throw std::logic_error("destinationIds is empty.");
+            }
+        }
+        catch (const std::logic_error&)
```
Comment (kbradl-uw) [inline HuntTheWumpusLib/Dungeon.cpp]: Laudable to check that it's empty and throw an error, but doing local recovery with a catch is not proper.  throw -> catch is meant to transfer control flow to a non-local context.

Don't do the try -catch here. Leave the throw, and handle that in e.g. HuntTheWumpus.cpp or a unit test.

**Review by kbradl-uw — COMMENTED** [review]
You have a good start on adding some detection points.

Three things to consider.

a) Prefer local recovery (if - do something) vs using try-catch for behaviors inside the same scope. try-throw-catch is more expensive than if-else.  Think about your control flow and if behaviors you're manifesting can be handled with logic instead of try-catch.  

b) For this assignment, think about interactions HuntTheWumpusLib has with external entities and what can it trust.  You've done one input validation in MakeMove. What about other interactions, particuarly through the Context?

c) Need to add unit tests for failure points identified in HuntTheWumpusLib. E.g. for your empty destinationIds in MakeMove, add a unit test that invokes MakeMove with empty destinationIds, and then verify proper behavior manifests.
